### PR TITLE
Fixed gRPC write path

### PR DESCRIPTION
**Pull request by singhravidutt**
GCS team have updated the checkpointing strategy for uplodObjects which resulted in bug for `gcs-connector` as we weren't able to get the updated write offset of file which is till getting uploaded. All upload object requests with size >40MB were resulting into failures.

Fix: Instead of waiting on `QueryWriteStatus` to provided the uploaded chunk now we mark completion of requestStream which result in hard checkpointing at gcs side. We do create subsequent requests to upload the remaining chunks.

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 1]
```diff
@@ -275,12 +275,27 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
+              storageStub
+                  .getChannel()
+                  .newCall(StorageGrpc.getWriteObjectMethod(), stub.getCallOptions());
+          writeObjectRequestStreamObserver =
+              ClientCalls.asyncClientStreamingCall(call, responseObserver);
+          requestStreamObserver =
+              watchdog.watch(
+                  call,
+                  writeObjectRequestStreamObserver,
+                  Duration.ofSeconds(channelOptions.getGrpcWriteMessageTimeoutMillis()));
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:292]: Should it be `Duration.ofMillis`?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 2]
```diff
@@ -488,6 +481,10 @@ public void beforeStart(
           ClientCallStreamObserver<WriteObjectRequest> clientCallStreamObserver) {
         clientCallStreamObserver.setOnReadyHandler(ready::countDown);
       }
+
+      public boolean isComplete() {
+        return done.getCount() != 0 ? false : true;
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:486]: Simplify to just `return done.getCount() == 0`

**Review by medb — APPROVED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 3]
```diff
@@ -275,12 +275,27 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
+          // reset streams
+          responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset);
+          call =
+              storageStub
+                  .getChannel()
+                  .newCall(StorageGrpc.getWriteObjectMethod(), stub.getCallOptions());
+          writeObjectRequestStreamObserver =
+              ClientCalls.asyncClientStreamingCall(call, responseObserver);
+          requestStreamObserver =
+              watchdog.watch(
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:289]: With this, the watchdog will be watching the previous call and the current call?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 4]
```diff
@@ -301,20 +316,24 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
                   : responseObserver.nonTransientError);
           break;
-        } else if (objectFinalized) {
+        }
+        if (objectFinalized
+            || requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
+          // close request stream if final request OR Map is full
           requestStreamObserver.onCompleted();
+          try {
+            responseObserver.done.await();
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:325]: Please add logs.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 5]
```diff
@@ -275,12 +275,27 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
+                  writeObjectRequestStreamObserver,
+                  Duration.ofSeconds(channelOptions.getGrpcWriteMessageTimeoutMillis()));
+        }
+
         WriteObjectRequest insertRequest = null;
         if (requestChunkMap.size() > 0 && requestChunkMap.lastKey() >= writeOffset) {
           insertRequest = getCachedRequest(requestChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
-        } else if (requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-          freeUpCommittedRequests(requestChunkMap, writeOffset);
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:283]: Will removing this affect the performance?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 6]
```diff
@@ -301,20 +316,24 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
           break;
-        } else if (objectFinalized) {
+        }
+        if (objectFinalized
+            || requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
+          // close request stream if final request OR Map is full
           requestStreamObserver.onCompleted();
+          try {
+            responseObserver.done.await();
+            requestChunkMap.clear();
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:326]: With this we are losing track of the offsets which are yet to be committed. What are the implications.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 7]
```diff
@@ -301,20 +316,24 @@ private WriteObjectResponse doResumableUpload() throws IOException {
                   ? responseObserver.transientError
                   : responseObserver.nonTransientError);
           break;
-        } else if (objectFinalized) {
+        }
+        if (objectFinalized
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:320]: Please add comment explaining WHY we are making this change. It is not very obvious from code why we are doing this.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 8]
```diff
@@ -488,6 +481,10 @@ public void beforeStart(
           ClientCallStreamObserver<WriteObjectRequest> clientCallStreamObserver) {
         clientCallStreamObserver.setOnReadyHandler(ready::countDown);
       }
+
+      public boolean isComplete() {
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:485]: can this be not public?

Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:485]: No, it has to be public as I need it to find the state of observer.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 3 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:289]: No previous stream will be removed from the open stream as soon as we have encountered either error in `responseObserver` or `onComplete`.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 6 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:326]: No, we are not we are cleaning up iff we have got a successful response from server (via ResponseObserver).

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 7 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:320]: Enhanced logging.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 5 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:283]: Removing this specifically will not affect the performance. Even it's not functional right now. But yes we do have correctness test plan scheduled for the new logic we are introducing.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 2 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:486]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 1 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:292]: Seems like it's been wrong from all along. Updated.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 3 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:289]: By call you mean stream? If so, No. Previous stream must have reached it's terminal state. We are making sure of it by checking state of responseObserver.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 9]
```diff
@@ -311,18 +310,27 @@ private WriteObjectResponse doResumableUpload() throws IOException {
...
             || requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-          // close request stream if final request OR Map is full
+          // We are closing request stream either
+          // 1. It's final request OR
+          // 2. If requestChunkMap is full -> to make sure that gcs checkpoint already sent chunks.
           requestStreamObserver.onCompleted();
           try {
             responseObserver.done.await();
+            if (responseObserver.hasTransientError() || responseObserver.hasNonTransientError()) {
+              // responseObserver
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:324]: remove?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 3 above)
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:288]: do we want to add an assert for the same here?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 7 above)
Comment (arunkumarchacko) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:315]: Most of the comments are covering WHAT we are doing. WHY we are doing the way we are is not very evident. Can we also add comments explaining the WHY?

**Review by arunkumarchacko — APPROVED** [review]

**Pull request merged**